dask_kubernetes/operator/daskcluster.py

```diff
@@ -0,0 +1,401 @@
+        )
+        scheduler_name = cluster["items"][0]["metadata"]["name"]
+        num_workers = spec["replicas"]
+        for i in range(num_workers):
+            data = build_worker_pod_spec(
+                name, namespace, spec.get("image"), uuid4().hex, scheduler_name
+            )
+            kopf.adopt(data)
+            worker_pod = await api.create_namespaced_pod(
+                namespace=namespace,
+                body=data,
+            )
+        logger.info(f"{num_workers} Worker pods in created in {namespace}")
+
+
+@kopf.on.update("daskworkergroup")
+async def daskworkergroup_update(spec, name, namespace, logger, **kwargs):
+    async with kubernetes.client.api_client.ApiClient() as api_client:
+        api = kubernetes.client.CoreV1Api(api_client)
+
+        scheduler = await kubernetes.client.CustomObjectsApi(
+            api_client
+        ).list_cluster_custom_object(
+            group="kubernetes.dask.org", version="v1", plural="daskclusters"
+        )
+        scheduler_name = scheduler["items"][0]["metadata"]["name"]
+        workers = await api.list_namespaced_pod(
+            namespace=namespace,
+            label_selector=f"dask.org/workergroup-name={name}",
+        )
+        current_workers = len(workers.items)
+        desired_workers = spec["replicas"]
+        workers_needed = desired_workers - current_workers
+        if workers_needed > 0:
+            for i in range(workers_needed):
+                data = build_worker_pod_spec(
+                    name, namespace, spec.get("image"), uuid4().hex, scheduler_name
+                )
+                kopf.adopt(data)
+                worker_pod = await api.create_namespaced_pod(
+                    namespace=namespace,
+                    body=data,
+                )
+            logger.info(f"Scaled worker group {name} up to {spec['replicas']} workers.")
+        if workers_needed < 0:
+            scheduler = SCHEDULER
+            worker_ids = await scheduler.workers_to_close(
+                n=-workers_needed, attribute="name"
+            )
+            logger.info(f"Workers to close: {worker_ids}")
+            for wid in worker_ids:
+                worker_pod = await api.delete_namespaced_pod(
+                    name=wid,
+                    namespace=namespace,
+                )
+            logger.info(
+                f"Scaled worker group {name} down to {spec['replicas']} workers."
+            )
+
+
+def patch_replicas(replicas):
+    patch = {"spec": {"replicas": replicas}}
+    json_patch = json.dumps(patch)
+    subprocess.check_output(
+        [
+            "kubectl",
+            "patch",
+            "daskworkergroup",
+            "default-worker-group",
+            "--patch",
+            str(json_patch),
+            "--type=merge",
+        ],
+        encoding="utf-8",
+    )
+
+
+@kopf.timer("daskworkergroup", interval=5.0)
+async def adapt(spec, name, namespace, logger, **kwargs):
```

Comment:
Need to handle multiple clusters.

Reply:
@philipp-sontag-by "We are now using a single thread that periodically iterates over all dask resources and makes the scaling decision for all of them."

Thank you for your comments! Can you tell me a little more about how you are doing this? I've run into a couple of the issues you listed.

Reply:
Basically we have a singleton custom resource that we have defined a `kopf.timer` for. Since this resource only exists once on the cluster we don't run into the worker problems. https://github.com/nolar/kopf/issues/642 has not been an issue for us since kopf introduced the retries. Alternatively you could just start a separate thread yourself in a `@kopf.configure` handler and do it in there.

In that `kopf.timer` handler we then iterate over all Dask resources on the cluster. We make use of [kopf indexing](https://kopf.readthedocs.io/en/stable/indexing/) here so that we don't have to query the API server in each iteration.
We then compute the worker allocation for each resource based on the total amount of resources available and the Dominant Resource Fairness algorithm. The allocation is the applied via a patch API call to our custom Dask resource.